### ISSUE TITLE
request for clarification: can ghci evaluate arbitrary expression

### ISSUE THREAD

**Issue by andrewufrank** [ndmitchell/ghcid#230]
I am new to use `vs code` and `ghcid` and I see that `All good` message, but I cannot any evaluation of expressions to work (like ghci typically does). I cannot see in the documentation if such a feature is not available or if there is an error in my operations or installation. 
I understand that I should be capable of having each compilation cycle a predetermined expression evaluated, but I cannot get this to work either. Could you give a complete `.ghci` file or the full command line to achieve this in the documentation? Thank you!

**Comment by ndmitchell** [comment]
Are you after `--test`? https://github.com/ndmitchell/ghcid/blob/master/README.md#i-want-to-run-my-tests-when-files-change. That has an example. Is that enough?

**Comment by andrewufrank** [comment]
yes, but the hint in the README is for a beginner not enough. I have in my .ghci file: 

```
:set -fwarn-unused-binds -fwarn-unused-imports -fno-warn-deprecations 
:set -fno-warn-unused-do-bind -fno-warn-missing-signatures
:set -isrc
:load test/ExampleTesting.hs
--test:=ExampleTesting
```

and would like to run the tests in the cabal

```
test-suite g2
  type: exitcode-stdio-1.0
  main-is: ExampleTesting.hs
  other-modules:
        ShakeStartTests
        Lib.FileMgt_test
```

but do not understand whether it is `g2` or `ExampleTesting.hs`. 

Further, when I run `ghcid` in the VS code terminal, it runs only once and does not repeat after each edit. I got this once running, but no recollection, how.
Sorry for the many 'stupid' questions - it is typically what one does initially not understand. Thank you for the help!

**Comment by willbush** [comment]
I found this because I'm having the same issue. 
I found [this](https://github.com/ndmitchell/ghcid/issues/208), which links to [here](https://www.parsonsmatt.org/2018/05/19/ghcid_for_the_win.html). However, I'm using `cabal new-repl` instead of `stack ghci` and perhaps it doesn't take the same type of arguments because I can't get it to work. I'll look into it more later, but perhaps you're using stack and this will work for you.

**Comment by andrewufrank** [comment]
Thank you - i saw one of the hints but did not understand the relevance then - thank you for pointing them out and add it to the record here. I will experiment more but hope that the documentation is improved (I cannot do it, not understanding what is going on and experience is, later, when one knows, one does not see the need anymore) - I saw another IDE going down for an inbalance between new development and nifty features and a lack of documentation of these. sad!

**Comment by willbush** [comment]
Alright I found `ghcid -c "cabal new-repl unit-tests" --run --restart=./src` works for me. Changes to the test files will reload, but changes to files in the `./src` directory causes `ghcid` to completely restart.


The `unit-tests` is the name of my `test-suite` in my cabal file (I'm just using that as an example).


The `--run` flag is equivalent to `-T "main"`, `--test=main`, `-r "main"`, `--run=main`. If one enters `cabal new-repl unit-tests` (or `ghci`, `stack ghci` with proper modules loaded) and types `main` in the repl, then that's basically what those flags are doing. Or at least that how it seems to me


I tried the `--reload` flag, but that doesn't work because I'm not loading modules from my `./src` directory. All my attempts to do so have failed with errors like the following:

```
cabal: Cannot open a repl for multiple components at once. The targets
'ticket-reservation' and 'unit-tests' refer to different components..

The reason for this limitation is that current versions of ghci do not support
loading multiple components as source. Load just one component and when you
make changes to a dependent component then quit and reload.
```

Clearly, from the blog I linked it's possible. Perhaps `stack ghci` supports loading multiple components. I haven't looked into that much further. However, for me just learning Haskell `--restart` is fast enough.

**Comment by ndmitchell** [comment]
OK, I'm struggling here to untangle the conversations. I don't think your problems have anything in common.

@andrewufrank `ghcid` knows nothing about Cabal. The `--test=XXX` is a command line flag to `ghcid`. I suggest you get your program working so `ghci` loads it. Then, in `ghci`, type a command - I suspect `main`, to run your tests. Once you figure out "I run ghci with flags and settings X" and then I type command Y, you can transfer that to `ghcid` with `--command=X` and `--test=Y`. I am happy to improve the docs, but not until I have figured out what issue you are having, since at the moment I couldn't improve the docs in a way that would fix your issue, since I don't understand it.

@willbush you are struggling to find a command line invocation that starts your project with the right things loaded? I suspect you have to ask the Cabal developers, or write your own .ghci file. If you are using `--reload=src` then you are almost certainly doing it wrong - the whole idea is that you load the files in `ghci` and then `ghci` knows how to reload.

**Comment by willbush** [comment]
Just to clarify my issue was that I wanted `ghcid` to reload and run tests after editing either a test file or a library file. I couldn't figure out how to get it to reload and run tests after editing a library file.

>the whole idea is that you load the files in ghci and then ghci knows how to reload.

Thanks, after I read that I worked on getting it working with just `ghci`. After some experimenting I found I needed to `:set -isrc` and `:load test/Spec.hs`.

```bash
nix-shell:~/ticket-reservation]$ ghci
GHCi, version 8.6.4: http://www.haskell.org/ghc/  :? for help
Loaded package environment from /home/will/ticket-reservation/.ghc.environment.x86_64-linux-8.6.4
[1 of 3] Compiling Types            ( src/Types.hs, interpreted )
[2 of 3] Compiling TicketSystem     ( src/TicketSystem.hs, interpreted )
[3 of 3] Compiling Spec             ( test/Spec.hs, interpreted )
Ok, three modules loaded.
Loaded GHCi configuration from /home/will/ticket-reservation/.ghci
*Spec> :set -isrc
*Spec> :l test/Spec.hs
[1 of 3] Compiling Types            ( src/Types.hs, interpreted )
[2 of 3] Compiling TicketSystem     ( src/TicketSystem.hs, interpreted )
[3 of 3] Compiling Spec             ( test/Spec.hs, interpreted )
Ok, three modules loaded.
*Spec>
```

Then I had everything needed loaded and entering `main` runs the tests. Now I have a `.ghci` file that looks like the following:

```
:set -fwarn-unused-binds -fwarn-unused-imports
:set -isrc
:load test/Spec.hs
```
And I initialize `ghcid` with: `ghcid --test=main`. Now from one terminal I get `ghcid` monitoring library and test code and running tests upon reload. **mission accomplished**

I think I was trying to hard to get it to work with `cabal new-repl` and didn't think to remove that layer of complexity.

I wouldn't mind contributing to documentation, if desired.

**Comment by andrewufrank** [comment]
I have a '.ghci' file 
```
:set -fno-warn-unused-binds -fnowarn-unused-imports
:set -fno-warn-missing-signatures
:set -isrc
:set -iUniform/Shake
:load Main
```
when I start `exec stack ghcid` I get a long list of warnings (including warnings about missing signatures) and output halts at an arbitrary location (but always the same for the same code) in the middle of the output - this is the same in vs code terminal and a ordinary linux terminal. The title says `138 warnings`.  This output is repeated when the code is changed and stops at the same location. The output is very long.
nothing happens - till I stop with ctrl-c ctrl-c.

**Comment by andrewufrank** [comment]
Same .ghci but the project has a main. now `stack exec ghcid` runs with some minimal output to `All good (7 modules, at ..), which is repeated when files are changed. How to I go to a interactive debugger?  If I start `ghci` in another terminal, I get the error message `could not find module xxx` (an import in the main) but it seems I have to say `stack exec ghci`.
The difficulty is to understand the interaction between stack or cabal and ghci and ghcid. 
Summary:
- stack exec ghcid -> gives repeated info for compilation errors
- stack exec ghci -> gives a debugger
Now I have only to find how to set the directories to find the modules which I need the debugger for.
Pure convenience: is there a way to avoid typing `:r` each time in ghci? ghcid does automatically reload.

Thank you for your efforts - the IDE is very important and your contribution is very useful!

**Comment by andrewufrank** [comment]
Unfortunately, when I add an import in the main (which then re-imports a long chain of other imports), I am again in the situation described before, where the output stops somewhere and nothing can be done anymore. Is this related to the interpretative nature of ghci and just too many warnings? (the code compiles with stack build)? 
I assume this is not a `ghcid` issue, except for an instructive error message and not just stopping.

**Comment by andrewufrank** [comment]
One of the issues was that I was (automatically ghc produced) importing `Paths_MODULENAME` which is not inserted in the list of `:set -i` produced by stack. It can be added with 

`:set -i.stack-work/dist/x86_64-linux-tinfo6/Cabal-2.4.0.1/build/autogen/`

and then `stack exec ghci` loads without error. This is not a `ghcid` problem, but one of cooperation of ghc, cabal, stack and ghci.

Still `ghcid` stops in the middle, for which I start a new issue.

**Comment by ndmitchell** [comment]
@willbush - documentation improvements most welcome! You tell me what/where you think needs clarifying.

@andrewufrank thanks, I'll take a look at the new issue. In general the output of `--verbose` in a github gist is very useful for tracking down things like stopping loading.
